v1.7 backports 2020-10-02

christarazi opened the pull request:
* #13319 -- Automatic rate limiting of endpoint API calls (@tgraf)
  * Note:
     * Added 3 new commits from the upstream PR; one for vendoring in an update to `golang.org/x/time`, another for the `--enable-api-rate-limit` flag, and subsequent Helm chart change for the flag
     * For other conflict resolutions, see commit msgs with `Backporter's notes`

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 13319; do contrib/backporting/set-labels.py $pr done 1.7; done
```

tgraf committed:
rate: Add API rate limiting system

[ upstream commit 3141e6581e02a895d9f7ed93f77850cbc48b9b9d ]

The API rate limiting system is capable to enforce both rate limiting
and maximum parallel requests. Instead of enforcing static limits, the
system is capable to automatically adjust rate limits and allowed
parallel requests by comparing the provided estimated processing
duration with the mean processing duration observed.

Usage:

```
var requestLimiter = rate.NewAPILimiter("myRequest", rate.APILimiterParameters{
	rate.SkipInitial:      5,
	rate.RateLimit:        1.0, // 1 request/s
	rate.ParallelRequests: 2,
}, nil)

func myRequestHandler() error {
	req, err := requestLimiter.Wait(context.Background())
	if err != nil {
		// request timed out whie waiting
		return err
	}
	defer req.Done() // Signal that request has been processed

	// process request ....

	return nil
}
```

Configuration parameters:

 - EstimatedProcessingDuration time.Duration

   EstimatedProcessingDuration is the estimated duration an API call
   will take. This value is used if AutoAdjust is enabled to
   automatically adjust rate limits to stay as close as possible to the
   estimated processing duration.

 - AutoAdjust bool

   AutoAdjust enables automatic adjustment of the values
   ParallelRequests, RateLimit, and RateBurst in order to keep the mean
   processing duration close to EstimatedProcessingDuration

 - MeanOver int

   MeanOver is the number of entries to keep in order to calculate the
   mean processing and wait duration

 - ParallelRequests int

   ParallelRequests is the parallel requests allowed. If AutoAdjust is
   enabled, the value will adjust automatically.

 - MaxParallelRequests int

   MaxParallelRequests is the maximum parallel requests allowed. If
   AutoAdjust is enabled, then the ParalelRequests will never grow above
   MaxParallelRequests.

 - MinParallelRequests int

   MinParallelRequests is the minimum parallel requests allowed. If
   AutoAdjust is enabled, then the ParallelRequests will never fall
   below MinParallelRequests.

 - RateLimit rate.Limit

   RateLimit is the initial number of API requests allowed per second.
   If AutoAdjust is enabled, the value will adjust automatically.

 - RateBurst int

   RateBurst is the initial allowed burst of API requests allowed. If
   AutoAdjust is enabled, the value will adjust automatically.

 - MinWaitDuration time.Duration

   MinWaitDuration is the minimum time an API request always has to wait
   before the Wait() function returns an error.

 - MaxWaitDuration time.Duration

   MaxWaitDuration is the maximum time an API request is allowed to wait
   before the Wait() function returns an error.

 - Log bool

   Log enables info logging of processed API requests. This should only
   be used for low frequency API calls.

   Example:
   ```
   level="info" msg="Processing API request with rate limiter" maxWaitDuration=10ms name=foo parallelRequests=2 subsys=rate uuid=933267c5-01db-11eb-93bb-08002720ea43
   level="info" msg="API call has been processed" name=foo processingDuration=10.020573ms subsys=rate totalDuration=10.047051ms uuid=933265c7-01db-11eb-93bb-08002720ea43 waitDurationTotal="18.665µs"
   level=warning msg="Not processing API request. Wait duration for maximum parallel requests exceeds maximum" maxWaitDuration=10ms maxWaitDurationParallel=10ms name=foo parallelRequests=2 subsys=rate uuid=933269d2-01db-11eb-93bb-08002720ea43
   ```

 - DelayedAdjustmentFactor float64

   DelayedAdjustmentFactor is percentage of the AdjustmentFactor to be
   applied to RateBurst and MaxWaitDuration defined as a value between
   0.0..1.0. This is used to steer a slower reaction of the RateBurst
   and ParallelRequests compared to RateLimit.

 - SkipInitial int

   SkipInitial is the number of API calls to skip before applying rate
   limiting. This is useful to define a learning phase in the beginning
   to allow for auto adjustment before imposing wait durations on API
   calls.

 - MaxAdjustmentFactor float64

   MaxAdjustmentFactor is the maximum adjustment factor when AutoAdjust
   is enabled. Base values will not adjust more than by this factor.

The configuration of API rate limiters is typically provided as-code to
establish defaults. A string based configuration option can then be used
to adjust defaults. This allows to expose configuration of rate limiting
using a single option flag:

``` go
l, err = NewAPILimiterSet(map[string]string{
	"foo": "rate-limit:2/m,rate-burst:2",
}, map[string]APILimiterParameters{
	"foo": {
		RateLimit:  rate.Limit(1.0 / 60.0),
		AutoAdjust: true,
	},
}, nil)
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

tgraf committed:
agent: Add rate limiting to endpoint API calls

[ upstream commit 6333eaac4d602cb65dc6a7544767399d40e81714 ]

[ Backporter's notes:
                      * v1.7 tree doesn't have `daemon/cmd`, so renamed
                        upstream `daemon/cmd/api_limits.go` ->
                        `daemon/api_limits.go`. This includes changing
                        the file's package from cmd -> main.
                      * Small conflict with parameters to
                        `createEndpoint()` ]

Add default rate limiting for all endpoint related API calls with
automatic adjustment based on estimated processing duration.

Metrics are provided to monitor the rate limiting system:
```
cilium_api_limiter_adjustment_factor                  api_call="endpoint-create"                               0.695787
cilium_api_limiter_processed_requests_total           api_call="endpoint-create" outcome="success"             7.000000
cilium_api_limiter_processing_duration_seconds        api_call="endpoint-create" value="estimated"             2.000000
cilium_api_limiter_processing_duration_seconds        api_call="endpoint-create" value="mean"                  2.874443
cilium_api_limiter_rate_limit                         api_call="endpoint-create" value="burst"                 4.000000
cilium_api_limiter_rate_limit                         api_call="endpoint-create" value="limit"                 0.347894
cilium_api_limiter_requests_in_flight                 api_call="endpoint-create" value="in-flight"             0.000000
cilium_api_limiter_requests_in_flight                 api_call="endpoint-create" value="limit"                 0.000000
cilium_api_limiter_wait_duration_seconds              api_call="endpoint-create" value="max"                  15.000000
cilium_api_limiter_wait_duration_seconds              api_call="endpoint-create" value="mean"                  0.000000
cilium_api_limiter_wait_duration_seconds              api_call="endpoint-create" value="min"                   0.000000
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

tgraf committed:
doc: Document API rate limiting

[ upstream commit d315ec3a4b26cd77d2edb4e2f4b79a148d42d246 ]

[ Backporter's notes: `operations/metrics.rst doesn't exist in v1.7,
                      it's `configuration/metrics.rst. Moved the
                      upstream changes there to resolve conflicts. ]

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
vendor: Bump golang.org/x/time

This commit bumps the package to v0.0.0-20191024005414-555d28b269f0,
which brings in support for setting bursts. This is required for an
upcoming commit for api rate limiting.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
agent, rate: Add option to control api rate limit

This commit introduces `--enable-api-rate-limit` to enable the use of
api rate limiting.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on daemon/api_limits.go:
```diff
@@ -0,0 +1,124 @@
+type apiRateLimitingMetrics struct{}
+
+func (a *apiRateLimitingMetrics) ProcessedRequest(name string, v rate.MetricsValues) {
+	metrics.APILimiterProcessingDuration.WithLabelValues(name, "mean").Set(v.MeanProcessingDuration)
+	metrics.APILimiterProcessingDuration.WithLabelValues(name, "estimated").Set(v.EstimatedProcessingDuration)
+	metrics.APILimiterWaitDuration.WithLabelValues(name, "mean").Set(v.MeanWaitDuration)
+	metrics.APILimiterWaitDuration.WithLabelValues(name, "max").Set(v.MaxWaitDuration.Seconds())
+	metrics.APILimiterWaitDuration.WithLabelValues(name, "min").Set(v.MinWaitDuration.Seconds())
+	metrics.APILimiterRequestsInFlight.WithLabelValues(name, "in-flight").Set(float64(v.CurrentRequestsInFlight))
+	metrics.APILimiterRequestsInFlight.WithLabelValues(name, "limit").Set(float64(v.ParallelRequests))
+	metrics.APILimiterRateLimit.WithLabelValues(name, "limit").Set(float64(v.Limit))
+	metrics.APILimiterRateLimit.WithLabelValues(name, "burst").Set(float64(v.Burst))
+	metrics.APILimiterAdjustmentFactor.WithLabelValues(name).Set(v.AdjustmentFactor)
+	metrics.APILimiterWaitHistoryDuration.WithLabelValues(name).Observe(v.WaitDuration.Seconds())
+	metrics.APILimiterProcessedRequests.WithLabelValues(name, metrics.Error2Outcome(v.Error)).Inc()
```
I wonder whether we should skip exporting the `limit` metrics if the functionality is disabled, since no limits would be enforced..?

christarazi replied:
Good point, will add line to bail out if this flag is not enabled after CI runs.

christarazi committed:
install: Add enableAPIRateLimit option

Following the commit to add the flag to the agent, this commit adds it
to the Helm chart. Note, that the api rate limiting itself is not
configurable through the Helm chart. It must be done directly on the
agent.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/option/config.go:
```diff
@@ -2106,10 +2120,8 @@ func (c *DaemonConfig) Populate() {
-	if val := viper.GetInt(ConntrackGarbageCollectorIntervalDeprecated); val != 0 {
-		c.ConntrackGCInterval = time.Duration(val) * time.Second
-	} else {
-		c.ConntrackGCInterval = viper.GetDuration(ConntrackGCInterval)
```
:thinking: how did we end up dropping the conntrackGCInterval?

christarazi replied:
Thanks. Fixed

joestringer commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -348,3 +348,4 @@ data:
   enable-well-known-identities: "false"
 {{- end }}
   enable-remote-node-identity: {{ .Values.global.remoteNodeIdentity | quote }}
+  enable-api-rate-limit: {{ .Values.global.enableAPIRateLimit | quote }}
```
We should be adding this to `config` rather than `global`.

christarazi replied:
I think given that this is 1.7, the precedent here is `global`, so I'll stick with that.
